airsonic-main/src/main/java/org/airsonic/player/service/search/QueryFactory.java

```diff
@@ -1,231 +1,291 @@
+ along with Airsonic.  If not, see <http://www.gnu.org/licenses/>.
+
+ Copyright 2016 (C) Airsonic Authors
+ Based upon Subsonic, Copyright 2009 (C) Sindre Mehus
+ */
+
+package org.airsonic.player.service.search;
+
+import org.airsonic.player.domain.MediaFile.MediaType;
+import org.airsonic.player.domain.MusicFolder;
+import org.airsonic.player.domain.RandomSearchCriteria;
+import org.airsonic.player.domain.SearchCriteria;
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.document.IntPoint;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause.Occur;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.WildcardQuery;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import static org.springframework.util.ObjectUtils.isEmpty;
+
+/**
+ * Factory class of Lucene Query.
+ * This class is an extract of the functionality that was once part of SearchService.
+ * It is for maintainability and verification.
+ * Each corresponds to the SearchService method.
+ * The API syntax for query generation depends on the lucene version.
+ * verification with query grammar is possible.
+ * On the other hand, the generated queries are relatively small by version.
+ * Therefore, test cases of this class are useful for large version upgrades.
+ **/
+@Component
+public class QueryFactory {
+
+    private static final String ASTERISK = "*";
+
+    @Autowired
+    private AnalyzerFactory analyzerFactory;
+
+    private final Function<MusicFolder, Query> toFolderIdQuery = (folder) -> {
+        // Unanalyzed field
+        return new TermQuery(new Term(FieldNames.FOLDER_ID, folder.getId().toString()));
+    };
+
+    private final Function<MusicFolder, Query> toFolderPathQuery = (folder) -> {
+        // Unanalyzed field
+        return new TermQuery(new Term(FieldNames.FOLDER, folder.getPath().getPath()));
+    };
+
+    /*
+     *  XXX 3.x -> 8.x :
+     *  "SpanOr" has been changed to "Or".
+     *   - Path comparison is more appropriate with "Or".
+     *   - If "SpanOr" is maintained, the DOC design needs to be changed.
+     */
+    private final BiFunction<@NonNull Boolean, @NonNull List<MusicFolder>, @NonNull Query> toFolderQuery = (
+            isId3, folders) -> {
+        BooleanQuery.Builder mfQuery = new BooleanQuery.Builder();
+        folders.stream()
+        	.map(isId3 ? toFolderIdQuery : toFolderPathQuery)
```

Comment:
There's a tab/indent problem here.

Reply:
I revised that according to your advice.